rstudio/rstudio
# Implement Quarto publishing

### Intent

This change makes it possible to publish Quarto documents and websites from the editor as well as the Viewer pane. 

It also relaxes the requirement that R Markdown and Quarto files be self-contained in order to publish to RPubs; we now convert them to self-contained automatically if necessary.

### Approach

The "Quarto-ness" of a document or website is somewhat orthogonal from its other attributes, so the largest component of this change is just plumbing "is this Quarto?" flags so that we can enable Quarto-specific treatment where necessary.  Most of that treatment has to do with how we get metadata about Quarto content: from Quarto itself (via `quarto metadata`) or by reading Quarto project information. 

The Quarto flag is also, ultimately, forwarded to the rsconnect package, where it is included in the metadata delivered to the `deployApp` function and then written into the manifest as a hint to RStudio Connect.  

The change requires a couple of tweaks to the rsconnect and rmarkdown packages, so it temporarily embeds those packages until we can get those changes on CRAN.

### Automated Tests

See https://github.com/rstudio/rstudio-ide-automation/issues/131. 

### QA Notes

This change will require thorough testing of every combination you can think of in which Something is published to Somewhere via a Method. Here are the combinations to explore:

1. Publish via (a) the editor toolbar, (b) the Viewer pane, and (c) the separate preview window. 
2. Publish (a) an R Markdown doc, (b) a Shiny R Markdown doc, (c) an R Markdown website, (d) multiple R Markdown documents, (e) a Quarto doc, (f) a Shiny Quarto doc, (g) a Quarto website, (h) a Plumber API, (i) a Shiny application, (j) a plain Markdown file, etc. 
3. Publish (a) as code, or (b) as rendered content.
4. Publish to (a) RPubs, (b) shinyapps.io, or (c) RStudio Connect.
5. Publish via (a) a first time click on the Publish button or (b) replaying an existing publish by clicking Republish or choosing an existing deployment.

So for example you might test publishing (1) via the editor, (2) a Shiny R markdown doc, (3) as code, (4) to RStudio connect, and (5) via a Republish click. If you're unsure whether the behavior you're seeing is correct, compare with a slightly older build: there are many scenarios that are intentionally walled off due to lack of support (for example RPubs can't handle any Shiny related content). 

Also note that RStudio Connect can't render Quarto content from source yet, so while you can publish it, you won't see rendered copies. 

### Checklist
